lib/galaxy_test/api/test_workflows.py

```diff
@@ -1244,6 +1273,54 @@ def test_run_subworkflow_replacement_parameters(self):
             details = self.dataset_populator.get_history_dataset_details(history_id)
             assert details["name"] == "moocow suffix"
 
+    @skip_without_tool("create_2")
+    def test_placements_from_text_inputs(self):
+        with self.dataset_populator.test_history() as history_id:
+            run_def = """
+class: GalaxyWorkflow
+inputs: []
+steps:
+  create_2:
+    tool_id: create_2
+    state:
+      sleep_time: 0
+    outputs:
+      out_file1:
+        rename: "${replaceme} name"
+      out_file2:
+        rename: "${replaceme} name 2"
+test_data:
+  replacement_parameters:
+    replaceme: moocow
+"""
+
+            self._run_jobs(run_def, history_id=history_id)
+            details = self.dataset_populator.get_history_dataset_details(history_id)
+            assert details["name"] == "moocow name 2"
+
+            run_def = """
+class: GalaxyWorkflow
+inputs:
+  replaceme: text
+steps:
+  create_2:
+    tool_id: create_2
+    state:
+      sleep_time: 0
+    outputs:
+      out_file1:
+        rename: "${replaceme} name"
```

Comment:
I thought ``#{}`` was for step parameters and ``${}`` was for workflow parameters? Is this wrong?

Reply:
... of course, that explains why this didn't work for me when I tried it in the editor. Which also resolves my comment on execute.py.

Reply:
https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/jobs/actions/post.py#L135 right?

``${}`` would only respect legacy workflow parameters or modern step parameters (of which I've augmented the workflow run form to just render one) and ``#{}`` would only respect tool input names?

Or do you mean the old workflow run form ... I guess I didn't look into that but who is going to have a step parameter with the same name as a legacy parameter in the tool state without having used the tool? I mean... seems an odd corner case but I can fix the legacy tool form to look at step parameters like they are legacy parameters and I think that would solve this.

Reply:
I guess I could also read this as "what if a legacy replacement_dict" is passed to the API and a modern step parameter was passed. My response would be for backward compatibility the ``replacement_dict`` is still used but the UI shouldn't render the options and API consumers are going to much prefer the modern way of specifying these. I think it is fine right?

Reply:
No, I think this comment here is resolved with me mixing up tool step and workflow parameters. But I do think referencing new style parameters with the `${param}` syntax in the editor is unfortunate. Is there a reason why we can't use the same logic we use for tool step parameters ?

Reply:
I don't really know what modeling that connection gets us at all (from modern workflow parameter to PJA).

There is a fixed universe of modern, fixed workflow input parameters - we could easily build a UI to assist users in resolving them when writing PJA. From the backend perspective, I don't know what modeling that connection gets us - we have a dictionary of the inputs, we will probably always need it for legacy reasons, we can validate them, we have a distinct and unambiguous syntax.

This seems good enough and a simpler user and backend experience. We could be more formal in future... but this isn't worse than the legacy parameters right?

Explicit modeling of the tool inputs was really good - because the syntax was ambiguous and you can't reason at all about tool inputs without loading the complex tool - which was highly Galaxy specific. These PJA are no problem to reason about in gxformat2, etc.. well outside the context of Galaxy.

Reply:
So the advantages that I see is that you can see where your parameter is used and you can set a PJA parameter based on the output of a expression tool, like `compose_text_param` (haven't tried, maybe that works now too if you reference the output?). Seeing the parameters in use is quite helpful.

I agree that we have other means to potentially build a UI, but if that UI can be the same between tool input parameters and PJA inputs that would be easier for users.

> This seems good enough and a simpler user and backend experience

Not saying we need to implement a UI now, but I don't think `${}` is a good user experience, and i'd rather see less than more use of it in the UI. If we plan to keep this on the backend then I have no objection with the use here.

Reply:
I guess I could go into why I disagree on some levels but I'll instead just say I think there are two areas where we probably sort of agree in terms of this direction.

- This isn't all PJA, this is just renaming. Those big workflows where the input name is preserved across 25 steps are an anti-pattern and even though tags and collections and step parameters and workflow reports didn't quite kill them they did a good job and I think #10118 will be their death kneel 🎆🍾 . In all cases, displaying structure or preserved metadata is better than encoding it in the name. 
- We do need a way to define parameters to a step that don't corresponding 1-to-1 with tool or explicit subworkflow inputs for CWL - and way to define expressions that map those to the underlying inputs - and that should include PJA when mapped to Galaxy. These are fairly fundamental to CWL and are useful I think.